AppCenterData/AppCenterData/Internal/LocalStore/MSDBDocumentStore.m

```diff
@@ -189,6 +191,108 @@ - (MSDocumentWrapper *)readWithToken:(MSTokenResult *)token documentId:(NSString
+
+  // Get effective device time to live.
+  NSInteger deviceTimeToLive = baseOptions ? baseOptions.deviceTimeToLive : kMSDataTimeToLiveDefault;
+
+  // Execute the query.
+  NSString *tableName = [MSDBDocumentStore tableNameForPartition:token.partition];
+  NSString *selectionQuery =
+      [NSString stringWithFormat:@"SELECT * FROM \"%@\" WHERE \"%@\" = \"%@\"", tableName, kMSPartitionColumnName, token.partition];
+  NSArray *listResult = [self.dbStorage executeSelectionQuery:selectionQuery];
+
+  // Return an error if no documents were found.
+  if (listResult.count == 0) {
+    NSString *errorMessage = [NSString stringWithFormat:@"Unable to find any document in local store for partition '%@'", token.partition];
+    MSLogWarning([MSData logTag], @"%@", errorMessage);
+
+    // return an empty list
+    MSPaginatedDocuments *documents = [[MSPaginatedDocuments alloc] initWithPage:[[MSPage alloc] initWithItems:localListItems]
+                                                                       partition:partition
+                                                                    documentType:documentType
+                                                                deviceTimeToLive:baseOptions.deviceTimeToLive
+                                                               continuationToken:nil];
+    return documents;
+  }
+
+  // Parse the documents.
+  for (id documentRow in listResult) {
+
+    // If an expired document is found exclude it from the list and delete in from the local store.
+    long expirationTime = [(NSNumber *)(documentRow[self.expirationTimeColumnIndex]) longValue];
+    NSString *documentId = documentRow[self.documentIdColumnIndex];
+    if (expirationTime != kMSDataTimeToLiveInfinite) {
+      NSDate *expirationDate = [NSDate dateWithTimeIntervalSince1970:expirationTime];
+      NSDate *currentDate = [NSDate date];
+      if (expirationDate && [expirationDate laterDate:currentDate] == currentDate) {
+
+        NSString *errorMessage =
+            [NSString stringWithFormat:@"Local document for partition '%@' and document ID '%@' expired at %@, discarding it",
+                                       token.partition, documentId, expirationDate];
+        MSLogWarning([MSData logTag], @"%@", errorMessage);
+
+        // Delete the local document when found to be expired
+        [self deleteWithToken:token documentId:documentId];
+      } else if ([kMSPendingOperationDelete isEqualToString:documentRow[self.pendingOperationColumnIndex]]) {
+
+        // Ignore document, if the pending operation is found to be Delete
+        NSString *message = [NSString
+            stringWithFormat:
+                @"Local document pending deletion in local storage for partition '%@' and document ID '%@', excluding from the list",
+                token.partition, documentId];
+        MSLogError([MSData logTag], @"%@", message);
+      } else {
+
+        // Deserialize document.
```

Comment:
This block and the two following it (Deserialize, "if operation NSNull", and "if etag NSNull") can be factored out of this method and `readWithToken`

Reply:
@smithab Unresolving... Is there a reason these can't or shouldn't be factored out?  I see there's a bit of a difference with the document row, but at the very least, the "operation == NSNull" and "ETag == NSNull" blocks should be able to be factored out

Reply:
Fixed it.